internal/os/iscsi/api.go

```diff
@@ -0,0 +1,192 @@
+	// ConvertTo-Json is not part of the pipeline because powershell converts an
+	// array with one element to a single element
+	cmdLine := fmt.Sprintf(
+		`ConvertTo-Json -InputObject @(Get-IscsiTargetPortal -TargetPortalAddress ` +
+			`${Env:iscsi_tp_address} -TargetPortalPortNumber ${Env:iscsi_tp_port} | ` +
+			`Get-IscsiTarget | Select-Object -ExpandProperty NodeAddress)`)
+	cmd := exec.Command("powershell.exe", "/c", cmdLine)
+	cmd.Env = append(os.Environ(),
+		fmt.Sprintf("iscsi_tp_address=%s", portal.Address),
+		fmt.Sprintf("iscsi_tp_port=%d", portal.Port))
+
+	out, err := cmd.CombinedOutput()
+	if err != nil {
+		return nil, fmt.Errorf("error discovering target portal. cmd: %s, output: %s, err: %w", cmdLine, string(out), err)
+	}
+
+	var iqns []string
+	err = json.Unmarshal(out, &iqns)
+	if err != nil {
+		return nil, fmt.Errorf("failed parsing iqn list. cmd: %s output: %s, err: %w", cmdLine, string(out), err)
+	}
+
+	return iqns, nil
+}
+
+func (APIImplementor) ListTargetPortals() ([]TargetPortal, error) {
+	cmdLine := fmt.Sprintf(
+		`ConvertTo-Json -InputObject @(Get-IscsiTargetPortal | ` +
+			`Select-Object TargetPortalAddress, TargetPortalPortNumber)`)
+
+	cmd := exec.Command("powershell.exe", "/c", cmdLine)
+	out, err := cmd.CombinedOutput()
+	if err != nil {
+		return nil, fmt.Errorf("error listing target portals. cmd %s, output: %s, err: %w", cmdLine, string(out), err)
+	}
+
+	var portals []TargetPortal
+	err = json.Unmarshal(out, &portals)
+	if err != nil {
+		return nil, fmt.Errorf("failed parsing target portal list. cmd: %s output: %s, err: %w", cmdLine, string(out), err)
+	}
+
+	return portals, nil
+}
+
+func (APIImplementor) RemoveTargetPortal(portal *TargetPortal) error {
+	cmdLine := fmt.Sprintf(
+		`Get-IscsiTargetPortal -TargetPortalAddress ${Env:iscsi_tp_address} ` +
+			`-TargetPortalPortNumber ${Env:iscsi_tp_port} | Remove-IscsiTargetPortal ` +
+			`-Confirm:$false`)
+
+	cmd := exec.Command("powershell.exe", "/c", cmdLine)
+	cmd.Env = append(os.Environ(),
+		fmt.Sprintf("iscsi_tp_address=%s", portal.Address),
+		fmt.Sprintf("iscsi_tp_port=%d", portal.Port))
+
+	out, err := cmd.CombinedOutput()
+	if err != nil {
+		return fmt.Errorf("error removing target portal. cmd %s, output: %s, err: %w", cmdLine, string(out), err)
+	}
+
+	return nil
+}
+
+func (APIImplementor) ConnectTarget(portal *TargetPortal, iqn string, isMultipath bool,
+	authType string, chapUser string, chapSecret string) error {
+	// Not using InputObject as Connect-IscsiTarget's InputObject does not work.
+	// This is due to being a static WMI method together with a bug in the
+	// powershell version of the API.
+	cmdLine := fmt.Sprintf(
+		`Connect-IscsiTarget -TargetPortalAddress ${Env:iscsi_tp_address}` +
+			` -TargetPortalPortNumber ${Env:iscsi_tp_port} -NodeAddress ${Env:iscsi_target_iqn}` +
+			` -AuthenticationType ${Env:iscsi_auth_type}` +
+			` -IsMultipathEnabled $([System.Convert]::ToBoolean(${Env:iscsi_is_multipath}))`)
```

Comment:
If `IsMultipathEnabled` is `true` then `Connect-IscsiTarget` will need to be invoked with `InitiatorPortalAddress` set to the IP address of each network adapter that wants to participate in the multi-path configuration. I think it will probably be better to block specification of `IsMultipathEnabled` for now?

Reply:
I didn't consider the multipath scenario where you use two initiators to the same target, what I was thinking about was the other direction
e.g. I have one NIC that connects to two IPs of the storage array,  each one exposing the same IQN.
This scenario will work with the current API
I guess this is more of a scenario for high availability rather than performance, and it was at least what we used in a product I've worked on (although it was far from the normal storage product)

WDYT? The multipath option here was added really quickly just to give an option for users with an already well configured node to use multipath. Full support will probably require more effort anyway, so I can remove the multipath bool and save it for a later version

Reply:
Interesting - most production multipath configurations I have seen involve redundant paths all the way from the initiator NICs to the target.

I would say let’s remove the multipath option for now in v1alpha1 and add it later if we hear requests/feedback around it.

Reply:
Once the multipath option is removed for now, I will lgtm/approve the PR. Thanks for getting basic iSCSI support going!

Reply:
Thanks! 
Removed the multipath option